alacritty/src/display/content.rs

```diff
@@ -47,21 +47,45 @@ impl<'a> RenderableContent<'a> {
         let focused_match = search_state.focused_match();
         let terminal_content = term.renderable_content();
 
+        let mut has_preedit_cursor_pos = false;
+        let mut preedit_cursor_delta: usize = 0;
+        let mut preedit_text_length: usize = 0;
+
+        match display.ime.preedit() {
+            None => {}
+            Some(the_preedit) => {
+                match the_preedit.cursor_byte_offset {
+                    None => {}
+                    Some(retriedved_offset) => {
+                        has_preedit_cursor_pos = true;
+                        preedit_cursor_delta = retriedved_offset;
+                        preedit_text_length = the_preedit.text.chars().count();
+                    }
+                }
+            }
+        }
+
         // Find terminal cursor shape.
         let cursor_shape = if terminal_content.cursor.shape == CursorShape::Hidden
             || display.cursor_hidden
             || search_state.regex().is_some()
-            || display.ime.preedit().is_some()
+            || (display.ime.preedit().is_some() && !has_preedit_cursor_pos)
         {
             CursorShape::Hidden
         } else if !term.is_focused && config.cursor.unfocused_hollow {
             CursorShape::HollowBlock
+        } else if has_preedit_cursor_pos {
+            CursorShape::Beam
```

Comment:
Shouldn't we just fall back to the normal text cursor? Not sure why we're forcing the beam?

Reply:
Inline Composition Buffer (ICB) is already underlined. Allowing normal text cursor can cause confusion if the user-specified underline cursor style overlaps the ICB underline.